flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -533,7 +541,7 @@ public ChangelogState getExistingState(String name, BackendStateType type)
         ChangelogState state;
         switch (type) {
             case KEY_VALUE:
-                state = (ChangelogState) keyValueStatesByName.get(name);
+                state = changelogStates.get(name);
```

Comment:
If I understand correctly, the `changelogStates` contains state which is already TTL state if previous TTL is enabled and having the TTL serializer?

I think this logic is a bit complex and we should better add some descriptions in the java doc. Moreover, how about rename the name `getExistingState` to like `getRestoredDelegatedState` to make the semantics more clean.

Reply:
If TTL is/was enabled, then all states will use TTL serializers. The difference between `changelogStates` and `keyValueStatesByName` is that the former does NOT wrap values into TTL values.
On recovery, values read from the changelog are TTL values. So using the latter would cause double-wrapping.

I agree, this needs to be documented.

As for the method, it can also return new non-restored state. So I'm going to rename it to `getExistingStateForRecovery`.  WDYT?

P.S.: we probably need to rethink TTL later, maybe we don't need to write TTL to changelog. But that means changing the semantics. And then we probably would change it for all states to be consistent (FLINK-9661).

Reply:
For now, `getExistingState` would only be used during recovery, will it be used in other places?

Reply:
Yes, it's only used during the recovery for now.